### PR TITLE
Monkeypatch each iframe

### PR DESCRIPTION
**Pull request by eoghanmurray**
observers which worked by monkeypatching built in objects were not working within iFrames as these have their own clean window object which doesn't include the iframe

### CODE REVIEW

**Comment by Yuyz0112** [comment]
Also found these bugs, good job 👍 . Going to review in the coming holiday.

File: packages/rrweb/src/record/observer.ts [hunk 1]
```diff
@@ -546,23 +547,23 @@ function initStyleSheetObserver(
   };
 
   const supportedNestedCSSRuleTypes: {
-    [key: string]: GroupingCSSRuleTypes;
+    [key: string]: win.GroupingCSSRuleTypes;
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:550]: This is a type, no need to prefix that
```suggestion
    [key: string]: GroupingCSSRuleTypes;
```

File: packages/rrweb/src/record/observer.ts [hunk 2]
```diff
@@ -575,8 +576,8 @@ function initStyleSheetObserver(
 
   Object.entries(supportedNestedCSSRuleTypes).forEach(([typeKey, type]) => {
     unmodifiedFunctions[typeKey] = {
-      insertRule: (type as GroupingCSSRuleTypes).prototype.insertRule,
-      deleteRule: (type as GroupingCSSRuleTypes).prototype.deleteRule,
+      insertRule: (type as win.GroupingCSSRuleTypes).prototype.insertRule,
+      deleteRule: (type as win.GroupingCSSRuleTypes).prototype.deleteRule,
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:580]: Also types
```suggestion
      insertRule: (type as GroupingCSSRuleTypes).prototype.insertRule,
      deleteRule: (type as GroupingCSSRuleTypes).prototype.deleteRule,
```

File: packages/rrweb/src/record/observer.ts [hunk 3]
```diff
@@ -622,11 +623,12 @@ function initStyleSheetObserver(
...
   cb: styleDeclarationCallback,
+  win: Window,
   mirror: Mirror,
 ): listenerHandler {
-  const setProperty = CSSStyleDeclaration.prototype.setProperty;
-  CSSStyleDeclaration.prototype.setProperty = function (
-    this: CSSStyleDeclaration,
+  const setProperty = win.CSSStyleDeclaration.prototype.setProperty;
+  win.CSSStyleDeclaration.prototype.setProperty = function (
+    this: win.CSSStyleDeclaration,
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:631]: Also a type
```suggestion
    this: CSSStyleDeclaration,
```

File: packages/rrweb/src/record/observer.ts [hunk 4]
```diff
@@ -648,9 +650,9 @@ function initStyleDeclarationObserver(
     return setProperty.apply(this, arguments);
   };
 
-  const removeProperty = CSSStyleDeclaration.prototype.removeProperty;
-  CSSStyleDeclaration.prototype.removeProperty = function (
-    this: CSSStyleDeclaration,
+  const removeProperty = win.CSSStyleDeclaration.prototype.removeProperty;
+  win.CSSStyleDeclaration.prototype.removeProperty = function (
+    this: win.CSSStyleDeclaration,
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:655]: Also a type
```suggestion
    this: CSSStyleDeclaration,
```

File: packages/rrweb/src/record/observer.ts [hunk 5]
```diff
@@ -702,26 +704,27 @@ function initMediaInteractionObserver(
...
-  const props = Object.getOwnPropertyNames(CanvasRenderingContext2D.prototype);
+  const props = Object.getOwnPropertyNames(win.CanvasRenderingContext2D.prototype);
   const handlers: listenerHandler[] = [];
   for (const prop of props) {
     try {
       if (
-        typeof CanvasRenderingContext2D.prototype[
-          prop as keyof CanvasRenderingContext2D
+        typeof win.CanvasRenderingContext2D.prototype[
+          prop as keyof win.CanvasRenderingContext2D
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:717]: Type
```suggestion
          prop as keyof CanvasRenderingContext2D
```

File: packages/rrweb/src/record/observer.ts [hunk 6]
```diff
@@ -702,26 +704,27 @@ function initMediaInteractionObserver(
...
         continue;
       }
       const restoreHandler = patch(
-        CanvasRenderingContext2D.prototype,
+        win.CanvasRenderingContext2D.prototype,
         prop,
         function (original) {
           return function (
-            this: CanvasRenderingContext2D,
+            this: win.CanvasRenderingContext2D,
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:727]: Type
```suggestion
            this: CanvasRenderingContext2D,
```

File: packages/rrweb/src/record/observer.ts [hunk 7]
```diff
@@ -757,8 +760,8 @@ function initCanvasMutationObserver(
       );
       handlers.push(restoreHandler);
     } catch {
-      const hookHandler = hookSetter<CanvasRenderingContext2D>(
-        CanvasRenderingContext2D.prototype,
+      const hookHandler = hookSetter<win.CanvasRenderingContext2D>(
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:763]: Also type
```suggestion
      const hookHandler = hookSetter<CanvasRenderingContext2D>(
```

File: packages/rrweb/src/record/observer.ts [hunk 8]
```diff
@@ -971,17 +974,26 @@ export function initObservers(
...
+  const curr_window = o.doc.defaultView,  // basically document.window
+
   const styleSheetObserver = initStyleSheetObserver(
     o.styleSheetRuleCb,
+    curr_window,
     o.mirror,
   );
   const styleDeclarationObserver = initStyleDeclarationObserver(
     o.styleDeclarationCb,
+    curr_window,
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:987]: JavaScript conventions
```suggestion
  const currentWindow = o.doc.defaultView,  // basically document.window

  const styleSheetObserver = initStyleSheetObserver(
    o.styleSheetRuleCb,
    currentWindow,
    o.mirror,
  );
  const styleDeclarationObserver = initStyleDeclarationObserver(
    o.styleDeclarationCb,
    currentWindow,
```

File: packages/rrweb/src/record/observer.ts [hunk 9]
```diff
@@ -971,17 +974,26 @@ export function initObservers(
...
     o.mirror,
   );
   const canvasMutationObserver = o.recordCanvas
-    ? initCanvasMutationObserver(o.canvasMutationCb, o.blockClass, o.mirror)
-    : () => {};
+    ? initCanvasMutationObserver(
+      o.canvasMutationCb,
+      curr_window,
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:993]:
```suggestion
      currentWindow,
```

**Review by Juice10 — CHANGES_REQUESTED** [review]

**Comment by eoghanmurray** [comment]
Working on the typings suggestions from Justin, thanks @Juice10!

File: packages/rrweb/src/record/observer.ts (hunk 8 above)
Comment (eoghanmurray) [reply packages/rrweb/src/record/observer.ts]: Thanks! My bad.

**Comment by eoghanmurray** [comment]
Thanks @Juice10 I think all the review suggestions have been incorporated with the last force-push

**Comment by Yuyz0112** [comment]
LGTM, and I will send a PR try to reduce the usage of `(win as any)`

**Pull request merged**